compiler/rustc_passes/src/dead.rs

```diff
@@ -44,16 +44,55 @@ fn should_explore(tcx: TyCtxt<'_>, def_id: LocalDefId) -> bool {
     )
 }
 
+fn struct_all_fields_are_public(tcx: TyCtxt<'_>, id: DefId) -> bool {
+    // treat PhantomData and positional ZST as public,
+    // we don't want to lint types which only have them,
+    // cause it's a common way to use such types to check things like well-formedness
+    tcx.adt_def(id).all_fields().all(|field| {
+        let field_type = tcx.type_of(field.did).instantiate_identity();
+        if field_type.is_phantom_data() {
+            return true;
+        }
+        let is_positional = field.name.as_str().starts_with(|c: char| c.is_ascii_digit());
+        if is_positional
+            && tcx
+                .layout_of(tcx.param_env(field.did).and(field_type))
+                .map_or(true, |layout| layout.is_zst())
+        {
+            return true;
+        }
+        field.vis.is_public()
+    })
+}
+
+/// pub struct whose fields are all public
+fn ty_ref_to_pure_pub_struct(tcx: TyCtxt<'_>, ty: &hir::Ty<'_>) -> bool {
+    if let TyKind::Path(hir::QPath::Resolved(_, path)) = ty.kind
+        && let Res::Def(def_kind, def_id) = path.res
+        && def_id.is_local()
+    {
+        return match def_kind {
+            DefKind::Enum | DefKind::Union => tcx.visibility(def_id).is_public(),
+            DefKind::Struct => {
+                tcx.visibility(def_id).is_public() && struct_all_fields_are_public(tcx, def_id)
+            }
+            _ => true,
+        };
+    }
+
+    true
+}
+
+/// pub struct but may have private fields
 fn ty_ref_to_pub_struct(tcx: TyCtxt<'_>, ty: &hir::Ty<'_>) -> bool {
```

Comment:
There is much duplicate between the two functions. I think we can have only one function which return two bool values, representing the visibility of the structure and all fields respectively.

Reply:
But `ty_ref_to_pure_pub_struct` does more logic than `ty_ref_to_pub_struct` to check the struct's fields.

Reply:
`ty_ref_to_pub_struct` is only called in two places. One of them will also call `ty_ref_to_pure_pub_struct` first. And it would have little impact to have some redundant functionality in another place. This will reduce duplication of code and prevent other developers from confusing the two.

Reply:
I pushed a new change for this ;)